modules/fluid_properties/src/fluidproperties/TemperaturePressureFunctionFluidProperties.C

```diff
@@ -238,85 +355,150 @@ TemperaturePressureFunctionFluidProperties::h_from_p_T(
 }
 
 Real
-TemperaturePressureFunctionFluidProperties::e_from_p_T(Real /* pressure */, Real temperature) const
-{
-  return _cv * temperature;
+TemperaturePressureFunctionFluidProperties::e_from_p_T(Real pressure, Real temperature) const
+{
+  if (_cv_is_constant)
+    return _e_ref + _cv * (temperature - _T_ref);
+  else
+  {
+    // Use Simpson's 3/8 rule to integrate cv
+    Real h = (temperature - _T_ref) / 3;
+    Real integral =
+        3. / 8. * h *
+        (cv_from_p_T(pressure, _T_ref) + 3 * cv_from_p_T(pressure, _T_ref + h) +
+         3 * cv_from_p_T(pressure, _T_ref + 2 * h) + cv_from_p_T(pressure, temperature));
+    return _e_ref + integral;
+  }
 }
 
 void
 TemperaturePressureFunctionFluidProperties::e_from_p_T(
     Real pressure, Real temperature, Real & e, Real & de_dp, Real & de_dT) const
 {
-  e = e_from_p_T(pressure, temperature);
-  de_dp = 0.0;
-  de_dT = _cv;
+  if (_cv_is_constant)
+  {
+    e = e_from_p_T(pressure, temperature);
+    de_dp = 0.0;
+    de_dT = _cv;
+  }
+  else
+  {
+    e = e_from_p_T(pressure, temperature);
+    // Use Simpson's 3/8 rule to integrate cv then take the derivative
+    Real h = (temperature - _T_ref) / 3;
+    Real cv0, dcv_dp0, dcv_dT0;
+    cv_from_p_T(pressure, _T_ref, cv0, dcv_dp0, dcv_dT0);
+    Real cv1, dcv_dp1, dcv_dT1;
+    cv_from_p_T(pressure, _T_ref + h, cv1, dcv_dp1, dcv_dT1);
+    Real cv2, dcv_dp2, dcv_dT2;
+    cv_from_p_T(pressure, _T_ref + 2 * h, cv2, dcv_dp2, dcv_dT2);
+    Real cv3, dcv_dp3, dcv_dT3;
+    cv_from_p_T(pressure, temperature, cv3, dcv_dp3, dcv_dT3);
+    de_dp = 3. / 8 * h * (dcv_dp0 + 3 * dcv_dp1 + 3 * dcv_dp2 + dcv_dp3);
+    // Consistency with how the energy is computed is more important here
+    de_dT =
+        1. / 8. * (cv0 + 3 * cv1 + 3 * cv2 + cv3) + 3. / 8 * h * (dcv_dT1 + 2 * dcv_dT2 + dcv_dT3);
+  }
 }
 
 Real
 TemperaturePressureFunctionFluidProperties::e_from_p_rho(Real p, Real rho) const
 {
-  Real temperature = T_from_p_rho(p, rho);
-  return _cv * temperature;
+  if (_cv_is_constant)
+  {
+    Real temperature = T_from_p_rho(p, rho);
+    return _e_ref + _cv * (temperature - _T_ref);
+  }
+  else
+  {
+    Real temperature = T_from_p_rho(p, rho);
+    return e_from_p_T(p, temperature);
+  }
 }
 
 Real
 TemperaturePressureFunctionFluidProperties::beta_from_p_T(Real pressure, Real temperature) const
 {
   Real rho, drho_dp, drho_dT;
   rho_from_p_T(pressure, temperature, rho, drho_dp, drho_dT);
-  return -drho_dT / rho;
+  return -drho_dp / rho;
```

Comment:
that's a bug fix

Reply:
I think it was right before.

Reply:
https://en.wikipedia.org/wiki/Compressibility

ugh. we use beta for what they call alpha
beta in our dosctring is "beta volumetric thermal expansion coefficient"

Reply:
Yeah that's pretty confusing. I see beta used for this coefficient sometimes, but alpha seems more common. At least the docstring appears correct and gives the units [1/K].

Reply:
I ll change the notation in a follow up PR